addons/coupon/models/coupon_program.py

```diff
@@ -0,0 +1,145 @@
+        ('on_next_order', 'Apply On Next Order')],
+        default='on_current_order', string="Applicability")
+    coupon_ids = fields.One2many('coupon.coupon', 'program_id', string="Generated Coupons", copy=False)
+    coupon_count = fields.Integer(compute='_compute_coupon_count')
+    company_id = fields.Many2one('res.company', string="Company", default=lambda self: self.env.company)
+    currency_id = fields.Many2one(string="Currency", related='company_id.currency_id', readonly=True)
+    validity_duration = fields.Integer(default=1,
+        help="Validity duration for a coupon after its generation")
+
+    @api.constrains('promo_code')
+    def _check_promo_code_constraint(self):
+        """ Program code must be unique """
+        for program in self.filtered(lambda p: p.promo_code):
+            domain = [('id', '!=', program.id), ('promo_code', '=', program.promo_code)]
+            if self.search(domain):
+                raise ValidationError(_('The program code must be unique!'))
+
+    @api.depends('coupon_ids')
+    def _compute_coupon_count(self):
+        coupon_data = self.env['coupon.coupon'].read_group([('program_id', 'in', self.ids)], ['program_id'], ['program_id'])
+        mapped_data = dict([(m['program_id'][0], m['program_id_count']) for m in coupon_data])
+        for program in self:
+            program.coupon_count = mapped_data.get(program.id, 0)
+
+    @api.onchange('promo_code_usage')
+    def _onchange_promo_code_usage(self):
+        if self.promo_code_usage == 'no_code_needed':
+            self.promo_code = False
+
+    @api.onchange('reward_product_id')
+    def _onchange_reward_product_id(self):
+        if self.reward_product_id:
+            self.reward_product_uom_id = self.reward_product_id.uom_id
+
+    @api.model
+    def create(self, vals):
+        program = super(CouponProgram, self).create(vals)
+        if not vals.get('discount_line_product_id', False):
+            discount_line_product_id = self.env['product.product'].create({
+                'name': program.reward_id.display_name,
+                'type': 'service',
+                'taxes_id': False,
+                'supplier_taxes_id': False,
+                'sale_ok': False,
+                'purchase_ok': False,
+                'lst_price': 0, #Do not set a high value to avoid issue with coupon code
+            })
+            program.write({'discount_line_product_id': discount_line_product_id.id})
+        return program
+
+    def write(self, vals):
+        res = super(CouponProgram, self).write(vals)
+        reward_fields = [
+            'reward_type', 'reward_product_id', 'discount_type', 'discount_percentage',
+            'discount_apply_on', 'discount_specific_product_ids', 'discount_fixed_amount'
+        ]
+        if any(field in reward_fields for field in vals):
+            self.mapped('discount_line_product_id').write({'name': self[0].reward_id.display_name})
+        return res
+
+    def unlink(self):
```

Comment:
we can also unlink the rule_id and the reward_id as having records of those two models seems to have no sense.

Probably we can remove coupons, or only some of them because in sale coupon they can be linked to sale_orders.